### DISCUSSION TITLE
Easy flashing without soldering (almost) using pogo pin adapter with CTS + DTR pins

### DISCUSSION THREAD

**Discussion by clowrey** [emporia-vue-local/esphome#53] [Show and tell]
Pictures show using a 6 pin pogo pin adapter +  CH340 serial converter allows you to flash with no pin holding etc as it has the required DTR + CTS pins. I have flashed about 7 of them like this so far. 

You should be able to figure out the pins following the colors ;) but it is as follows
Adapter : Emporia
DTR : IO0
CTS : EN
GND : GND
RX : RXD
TX : TXD
VCC : VCC_5V0

You do have to cut a surface mount trace and solder the jumper to convert the VCC pin on the serial adapter from 3.3V to 5V. Although it would probably work with 3.3V as the converter appears to have a 3.3V regulator. If you get the 7pin pogo adapter(I already had the 6 pin) you could try connect VCC to the VCC_3V3 and avoid changing the jumper on serial adapter to 5V. 

I don't even bother backing up the original firmware as I know I'm never going back..

Also if using the pogo pin adapter - the pads have allot of solder flux residue from the factory that will prevent it from working most likely (it got me for awhile) so you want to clean that up with fresh solder or scrape it off with something carefully.. You can see it still on the 3.3V pad on my image since I don't use that pad.

Another big "gotcha" I just want to remind everyone about is the antenna!! if you don't connect the external antenna it will not connect to your Wi-Fi :) so make sure you have that connected so you can verify your flash worked.. 

Parts:
Pogo pin adapter, get the 7p single: https://s.click.aliexpress.com/e/_DkK79Jd
This serial adapter has the CTS+DTR pins you want, most dont have both: https://s.click.aliexpress.com/e/_Dlx0Fzp
Male to female jumper wire: https://www.aliexpress.us/item/3256803468186263.html

![PXL_20221031_202748286](https://user-images.githubusercontent.com/6935928/199107918-2536b389-15d7-4a2f-b3bf-d68017f46c45.jpg)
![PXL_20221031_202738486](https://user-images.githubusercontent.com/6935928/199107942-945491a5-eb0c-4351-9714-df9c52e42c0e.jpg)
![PXL_20221031_202829957](https://user-images.githubusercontent.com/6935928/199107955-b9bbfade-96ae-4454-a660-38bfb2bded9d.jpg)
![PXL_20221031_202858870](https://user-images.githubusercontent.com/6935928/199107977-662b3750-b437-4e65-812c-e52a8de6830f.jpg)
![PXL_20221031_203423484](https://user-images.githubusercontent.com/6935928/199107990-dede622f-63e8-4739-a1c5-f39479411d12.jpg)

**Comment by airdrummingfool** [comment]
Thank you for this guide! It was very helpful. I wanted to add a note in case anyone runs into the same issues that I did.

My meter is REV-5, and when I wired up the pins the way you list them I was unable to access the ESP. I kept getting the following error:
`A fatal error occurred: Failed to connect to Espressif device: No serial data received.`
I read the project README a bit closer and realized that I needed to connect **`RTS`** to `EN` (instead of `CTS` to `EN`). Once I did that I was able to make a serial connection.

Additionally, I was getting the error `A fatal error occurred: Invalid head of packet (0xFF): Possible serial noise or corruption.` when trying to back up the original firmware. To fix this I just set the baud to 115200 (`--baud=115200`) and everything worked.

Thanks again!

**Reply by frankyj226** [reply]
Thanks for that CTS correction, this is the solution that finally worked for me on a rev5 board:

DTR : IO0
**RTS** : EN
GND : GND
RX : RXD
TX : TXD
VCC : VCC_5V0

I used [this adaptor.](https://www.amazon.com/dp/B075N82CDL?psc=1&ref=ppx_yo2ov_dt_b_product_details)

**Reply by wayneconnolly** [reply]
Hi, Great guide and thanks for the links.

**Reply by wayneconnolly** [reply]
I can confirm that the 7 pin pogo and the 3v connection work perfectly and its a green light.
I bought all the parts that were recommended and I am glad that I did. Thank you!

I have a REV-5 and from the `IO0` pogo pin, I simply held it onto the CPU chip when I plugged in the USB. It put it into bootloader mode straight away.

![emporia-vue-2](https://github.com/emporia-vue-local/esphome/assets/4142087/fef642af-80d3-422a-83f1-6507fce77a8b)

I only ran 2 commands and its all working.

`esptool.py read_flash 0 0x800000 flash_contents.bin`

`esphome run vue-utility-solar.yaml` (I had to reboot the device and put into bootloader again as I let the `IO0` off the CPU about 20 seconds after its started.

Thanks everyone!!

**Reply by clowrey** [reply]
Picture?

**Reply by wayneconnolly** [reply]
Updated and added pic

**Reply by clowrey** [reply]
Glad you got it working!

**Reply by Cyphon00** [reply]
Thanks for this guide/info. :)

I also had to use the IO0 being grounded to the chassis trick.

I couldn't be bothered ordering/waiting for pogo pins, so I held the programmer (6 pin header) to the correct pins for ~2 mins while the flash happened.
This was a bit of a pain tbh, but gladly I only had to do this twice - backup stock firmware and flash ESPhome.

The most helpful part to know whether I was in programming mode (or just holding the programmer and waiting blindly) was that the vue2 will beep on normal startup, which means it's NOT in bootloader mode.
If you get the connection timing correct (by ensuring the IO0 is grounded before VCC is applied), then the green light comes on and there's no beep!

Hope that helps.

**Comment by wayneconnolly** [comment]
Nice

Sent from Nine
________________________________

**Comment by dalklein** [comment]
Thanks everyone!!  Thanks flaviut!   

I needed to hold io0 to ground, while powering it up.  With the serial basic ch340 adapter, with CTS,DTR connected as shown in this thread, it wouldn't flash.  Removed those two wires, and only hold io0 to ground while powering up.  Then it took me a few times to realize, don't keep holding io0 to ground the whole time, which seemed to prevent it from flashing.

In any case, it works and spits out on MQTT, I love it!  I am a tiny person, with tiny microcontrollers, standing on the shoulders of giants!  Thank you!

dklein@TexasToast:/homez/dklein/Documents/Projects/uC/emporia_vue$ mosquitto_sub -h 192.168.15.15 -t gridvue/power/sensor/ashpgrid/state -v
gridvue/power/sensor/ashpgrid/state 15.3
gridvue/power/sensor/ashpgrid/state 15.8

Now I can have my new tiny 8266 heatpump meter (measuring water temp in/out and flow rate), also consume this electrical power signal and provide COP back over MQTT!

**Comment by epickayak** [comment]
This pogo setup also worked for me.  One key point for me was noted by [alklein](https://github.com/dalklein) above.  It would not connect/flash for me when I had the ground jumper also "connected" to the GPIO pin.  I needed to disconnect it immediately after boot which then allowed it to flash.  Worked great and gives me time to keep practicing soldering

**Comment by CJOWood** [comment]
Haha, I wish the pins lined up like that for the Vue 3! I shared this in another discussion... but I had to go a different way...

> Received my Vue 3 a few days ago and got around to flashing it last night... Didn't feel like soldering and did not have needle probes or a bdm frame... So I made do. I am equally proud and disgusted with myself :).
> ![PXL_20241111_051934691.jpg](https://github.com/user-attachments/assets/46afcbde-072b-4e82-8dc8-d7436c62674f)

**Reply by matt-avila** [reply]
Thats butt-ugly, but it works! I've done worse :-)

**Comment by chicagoandy** [comment]
Can someone point me to the correct USB to Serial adapter? I am having trouble finding CH340 with DTR and CTS pins.

**Reply by flaviut** [reply]
I guess "CTS" was probably a common labeling mistake on these adapters back when this guide was written? RTS is the correct pin, according to the official ESP32 docs:

> esptool.py resets ESP32 automatically by asserting DTR and RTS control lines of the USB-to-UART bridge, i.e., FTDI or CP210x (for more information, see [Establish Serial Connection with ESP32](https://docs.espressif.com/projects/esp-idf/en/v5.1-beta1/esp32/get-started/establish-serial-connection.html)). The DTR and RTS control lines are in turn connected to GPIO0 and CHIP_PU (EN) pins of ESP32, thus changes in the voltage levels of DTR and RTS will boot ESP32 into Firmware Download mode. As an example, check the [schematic](https://dl.espressif.com/dl/schematics/esp32_devkitc_v4-sch-20180607a.pdf) for the ESP32 DevKitC development board.

It's relatively easy for me to find the right adapter, for example: https://www.amazon.com/dp/B0D8T31GX8/

Just look for one that has about 8 pins in the thumbnail, compared to the standard ones that just have the 4 RX TX VCC GND.
